### PR TITLE
[SE-0358][stdlib] Adopt primary associated types

### PR DESCRIPTION
**Pull request by lorentey**
This draft PR experiments with the adoption of the proposed new language feature for [lightweight same-type syntax for primary associated types][SE-0346] in the Swift Standard Library.

[SE-0346]: https://github.com/apple/swift-evolution/blob/main/proposals/0346-light-weight-same-type-syntax.md

### CODE REVIEW

File: stdlib/public/core/SIMDVector.swift [hunk 1]
```diff
@@ -71,11 +73,17 @@ public protocol SIMDScalar {
...
 /// A SIMD vector of a fixed number of elements.
-public protocol SIMD: SIMDStorage,
-                      Codable,
-                      Hashable,
-                      CustomStringConvertible,
-                      ExpressibleByArrayLiteral {
+public protocol SIMD<
+  // FIXME: This ought to refer to `SIMDStorage.Scalar`, not introduce a new
+  // associated type.
+  Scalar
```
Comment (slavapestov) [inline stdlib/public/core/SIMDVector.swift:79]: This is an ABI break

**Comment by lorentey** [comment]
This now reflects the latest revision of SE-0346, which made things a lot nicer!

File: stdlib/public/core/Codable.swift [hunk 2]
```diff
@@ -938,7 +938,7 @@ public struct KeyedEncodingContainer<K: CodingKey> :
 ///
 /// Decoders should provide types conforming to `UnkeyedDecodingContainer` for
 /// their format.
-public protocol KeyedDecodingContainerProtocol {
+public protocol KeyedDecodingContainerProtocol<Key> {
```
Comment (benrimmington) [inline stdlib/public/core/Codable.swift:941]: This was missed in the previous commit:

```suggestion
public protocol KeyedDecodingContainerProtocol {
```

**Comment by lorentey** [comment]
@swift-ci test

File: stdlib/public/core/LazySequence.swift [hunk 3]
```diff
@@ -128,7 +128,8 @@
 /// and discards the resulting array. Instead, use `reduce` for summing 
 /// operations, or `forEach` or a `for`-`in` loop for operations with side 
 /// effects.
-public protocol LazySequenceProtocol: Sequence {
+public protocol LazySequenceProtocol<Elements>: Sequence
+where Elements.Element == Element {
```
Comment (slavapestov) [inline stdlib/public/core/LazySequence.swift:132]: Did you mean to add this? The constraint is already on the associatedtype below

Comment (lorentey) [reply stdlib/public/core/LazySequence.swift:132]: 🤦‍♂️ No, this was left over from the previous version.

**Comment by lorentey** [comment]
(The failed builds are expected; this feature is currently gated on `-enable-parameterized-protocol-types`, which this PR does not enable for the stdlib yet.)

**Comment by lorentey** [comment]
@swift-ci test

**Comment by milseman** [comment]
Status of this work? This is blocking regex algorithms, which currently have an ABI of returning Array (i.e. force an eager and allocation)

**Comment by lorentey** [comment]
@swift-ci test

**Pull request merged**